pkg/cmd/checkendpoints/controller/updates_manager.go

```diff
@@ -0,0 +1,75 @@
+package controller
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/operatorcontrolplane/podnetworkconnectivitycheck/v1alpha1helpers"
+)
+
+// NewUpdatesManager returns a queue sorting UpdateManager with the specified sorting window.
+func NewUpdatesManager(window time.Duration) *updatesManager {
+	return &updatesManager{
+		window:       window,
+		sortingQueue: map[time.Time][]v1alpha1helpers.UpdateStatusFunc{},
+	}
+}
+
+// UpdatesManager manages a queue of updates. The lock must be obtained before
+// invoking any of the methods.
+type UpdatesManager interface {
+	sync.Locker
+	// Add an update to the queue.
+	Add(timestamp time.Time, updates ...v1alpha1helpers.UpdateStatusFunc)
+	// Return the contents of the queue.
+	Queue() []v1alpha1helpers.UpdateStatusFunc
+	// Delete the contents of the queue.
+	Clear()
+}
+
+// updateManage implements an UpdateManager sorts the incoming updates and queues updates
+// outside of a sorting window, anchored on one end by the latest update, for processing.
+type updatesManager struct {
+	sync.Mutex
+	window          time.Duration
+	sortingQueue    map[time.Time][]v1alpha1helpers.UpdateStatusFunc
+	processingQueue []v1alpha1helpers.UpdateStatusFunc
+	timestamps      []time.Time
+}
+
+// Add an update to the queue. There is a delay equal to the size of the sorting window before
+// updates are made available on the queue to allow for updates submitted out of order within
+// the sorting window to be sorted by timestamp.
+func (u *updatesManager) Add(timestamp time.Time, updates ...v1alpha1helpers.UpdateStatusFunc) {
+	u.sortingQueue[timestamp] = updates
+
+	u.timestamps = append(u.timestamps, timestamp)
+	sort.Slice(u.timestamps, func(i, j int) bool {
+		return u.timestamps[i].Before(u.timestamps[j])
+	})
+
+	latestTimestamp := u.timestamps[len(u.timestamps)-1]
+	tmp := u.timestamps[:0]
+	for _, timestamp := range u.timestamps {
```

Comment:
yeah, you need to comment this.  Looks like this an attempt to having a sliding window of results based on time. your window needs to be at least one second larger than the delay though.

Do you need to delay the events too?

Reply:
The window is initialized as `delay + check period`, so in this case 10s (conn timeout) + 1s (check period).

Reply:
See followup PR where I've tried to improve this further: https://github.com/openshift/cluster-kube-apiserver-operator/pull/924.